cockroachdb/cockroach
# kvserver: add rac2 send queue functionality test

This commit introduces two testing knobs:

```go
// OverrideBypassAdmitWaitForEval is used to override the behavior of
// WaitForEval. When bypass is set to true, WaitForEval will return
// immediately and return the waited value. Otherwise, when bypass is set
// to false, or unset, WaitForEval will behave normally.
OverrideBypassAdmitWaitForEval func() bool
// OverrideAlwaysRefreshSendStreamStats is used to override the behavior of
// the send stream stats refresh. When set to true, the send stream stats
// will always be refreshed on a HandleRaftEventRaftMuLocked call. Otherwise,
// when set to false, the default behavior will be used.
OverrideAlwaysRefreshSendStreamStats bool
```

Also:

- Thread send queue precise size adjustments via a method which also
  asserts that the size is consistent between the count and bytes.
- Change the logging in `replicaSendStream.admitRaftMuLocked` to log
  when some tokens were actually returned. The existing logging was very
  spammy and made it hard to debug.


---


The header was incorrectly `range_id` rather than `store_id` for some
`TestFlowControl.*` integration tests. Amend the header to be
`store_id`.

---

Introduce `TestFlowControlSendQueue`, which exercises:

1. Send queue formation via send token exhaustion
2. Send queue force-flushing via a node disconnect
3. Send queue formation prevention via a testing knob to skip
   wait-for-eval

More specifically, the test has the following structure:

```sql
-- This test exercises send queue formation, prevention and flushing.
-- The structure roughly follows:
--   Start with three voters on [n1,n2,n3], where n1 is the leader+leaseholder.
--   Large regular write -4 MiB.
--   Allow admission [n1,n2].
--   - Tokens should be returned for n1 and n2.
--   Block admission [n2].
--   Regular write -1 MiB.
--   - Shouldn't be blocked on wait-for-eval because of quorum [n1,n2].
--   - Metrics should reflect send queue formation on n3.
--   Stop n2.
--   Regular write -1 MiB.
--   - Blocks on wait-for-eval, however the test bypasses this instance.
--   - Metrics should reflect n3 being force flushed.
--   Allow admission [n1,n2,n3].
--   Start n2.
--   Add n4, n5, the voters now are [n1,n2,n3,n4,n5].
--   Block admission [n4,n5] (already blocked)
--   Regular write -4 MiB.
--   Regular write -1  MiB.
--   - Shouldn't be blocked on wait-for-eval because of quorum [n1,n2,n3]
--   - Metrics should reflect send queue formation on n4,n5.
--   Unblock admission [n4,n5].
--   - Wait for tokens to be returned.
--   Block admission [n2,n3,n4,n5].
--   Regular write -4 MiB.
--   Regular write -1  MiB.
--   - Blocks on wait-for-eval, however the test bypasses this instance.
--   - Metrics should reflect 2 streams being prevented from forming a send queue.
--   Allow admission [n1,n2,n3,n4,n5] (all).
--   Assert all tokens returned.
```

Part of: https://github.com/cockroachdb/cockroach/issues/132614
Release note: None
